### DISCUSSION TITLE
uv instead of pip?

### DISCUSSION THREAD

**Discussion by yajo** [nix-community/dream2nix#1019] [Ideas]
It's known that uv is faster than pip, and it locks in parallel. That's by itself a good reason to experiment as an alternative to creating d2n locks.

However, they finally ship a multiplatform lock format! See https://github.com/astral-sh/uv/issues/3347#issuecomment-2231387995. There's a whole project ongoing to stabilize that effort and the format.

Do you think it'd be worth it to invest some time on trying to replace pip by uv, or at least adding a new backend for python projects?

**Comment by phaer** [comment]
Yes, I personally just lock resources but am following uv with great interest for the same reason. @purepani also showed interest into giving that a try iirc.

**Reply by phaer** [reply]
Thanks for highlighting the lock format btw, that wasn't there when I had last looked at uv. 

https://github.com/astral-sh/packse/blob/main/uv.lock contains a nice example, all in all it locks a lot like the discussion in https://discuss.python.org/t/lock-files-again-but-this-time-w-sdists/46593.

I believe this is very interesting for the dream2nix use case as this locks sdists *and* wheels, meaning one could toggle between source builds and re-using wheels without the need to re-locking.

**Comment by purepani** [comment]
That's definitely something to add as a backend, but we should wait till the lock file format is stabilized: https://github.com/astral-sh/uv/issues/4893

**Reply by yajo** [reply]
The advantage of testing early is that we could influence its stabilization. Current state is quite good but I'm worried specically about editable installs. 

FWIW, when https://github.com/astral-sh/uv/issues/1442 is closed, there's a chance to use it as a drop-in replacement for pip here.

**Reply by phaer** [reply]
I agree that the lock-file stability wouldn't need to be a blocker for a first integration.

But looking at the latest release, it looks like the unstable lock flag has been removed from `uv pip compile` and is only supported by `uv lock` atm. The latter requires a pyproject.nix and doesn't allow arbitrary requirement strings. So the interface would probably resemble that of our pdm module.

**Reply by purepani** [reply]
In that case I might cook something after an exam I have in august.

**Reply by purepani** [reply]
As a side note, it looks like pdm is going to use the UV resolver soon, so we might not need to do much to get that speedup:
https://github.com/pdm-project/pdm/issues/2693

**Reply by purepani** [reply]
Looks like they stabilized the lockfile(or at least, it is ready for stabilization), but it looks like the nixpkgs version for uv is out of date due to requiring rust 1.80, which is currently in staging, so I'm just going to wait till staging is merged to do this.

**Reply by purepani** [reply]
> As a side note, it looks like pdm is going to use the UV resolver soon, so we might not need to do much to get that speedup: [pdm-project/pdm#2693](https://github.com/pdm-project/pdm/issues/2693)

Has been experimentally added! PR Here: https://github.com/nix-community/dream2nix/pull/1067

**Comment by DavHau** [comment]
Does uv support multi system lock files by now?

**Reply by purepani** [reply]
It does(though I haven't tried it myself yet).

**Comment by DavHau** [comment]
PDM works pretty well I think and we already have a module for it, so why switch to uv?

**Reply by purepani** [reply]
It's significantly faster at locking(not a huge project but a difference of 1.5 seconds and ~10 seconds on one of my projects) l, and i would imagine that being able to support multiple package managers would allow for increased adoption

Though it's probably better if that RFC for lock files comes into existence at some point.

**Reply by phaer** [reply]
The proposed lock file also has one major advantage for source builds: It locks build-requires.

See the new PEP at https://peps.python.org/pep-0751/#package-build-requires

**Reply by yajo** [reply]
FWIW uv is nice in the sense that it isn't a build backend. Instead, it's just a development tool that lets each library have its own backend and plays nice with all of them. 

It replaces `python -m venv .venv && ./.venv/bin/pip install -e .` basically.

PDM is a build backend with its own peculiarities. But if you prefer hatchling, setuptools, poetry, etc., it doesn't help AFAIK.

**Reply by purepani** [reply]
PDM has a backend, but you don't have to use it, and can use whatever you define in pyproject.
Plus, astral is developing their own experimental backend: https://github.com/astral-sh/uv/issues/3957

**Comment by yajo** [comment]
I discovered https://github.com/pyproject-nix/uv2nix. Would it make sense for both projects to join forces? cc @adisbladis

**Reply by purepani** [reply]
I think uv2nix should probably exist on it's own, but it's definitely worth trying to integrate uv in general.

what @adisbladis is doing, from what I understand, is creating a completely different build infrastructure with pyproject.nix, replacing the one in nixpkgs. So there are 2 different things that dream2nix could do(either separately or together).

1. Integrate uv for locking. This is easier and gains the benefits of having speedy locking(though right now we can get that by turning on the pdm experimental uv support).
2. Create a new `buildPythonPackage` module(or something of the sort), and use pyproject.nix's infrastructure to build it. uv2nix is an example of this. This gains the benefits of not having to deal with the nixpkgs python hook hacks(like PYTHONPATH leaking), since pyproject.nix uses normal venvs.

**Reply by adisbladis** [reply]
> Would it make sense for both projects to join forces?

Short answer: No.

Long answer:
No. I'm convinced the right way to go about building good package manager integrations is from the bottom up with a very deep level of understanding of both Nix & the build systems & packaging semantics involved.
Those will entirely shape your UX & abstractions.

I don't believe in an approach where you build tooling starting with a desired UX and build your way from there, neither do I believe in the NixOS module system applied in the context of packaging.
